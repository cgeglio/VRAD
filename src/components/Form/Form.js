import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './Form.scss';

class Form extends Component {
  constructor() {
    super();
    this.state = { name: '', email: '', purpose: '', error: false };
  }

  handleChange = event => {
    this.setState({ [event.target.name]: event.target.value, error: false });
  }

  checkInputs = event => {
    event.preventDefault();
    const formValues = Object.values(this.state);
    if (!formValues.includes('')) {
      this.addUser(event);
    } else {
      this.setState({ error: true });
    }
  }

  addUser = event => {
    event.preventDefault();
    const { name, email, purpose } = this.state;
    const { addUser } = this.props;
    addUser({ name, email, purpose });
    this.setState({ name: '', email: '', purpose: '', error: false });
  }


  render() {
    const { name, email, purpose, error } = this.state;
    return (
      <form>
        <input
          type="text"
          placeholder="Name..."
          name="name"
          value={name}
          onChange={this.handleChange}
        />
        <input
          type="text"
          placeholder="Email..."
          name="email"
          value={email}
          onChange={this.handleChange}
        />
        <input
          type="text"
          placeholder="Purpose..."
          name="purpose"
          value={purpose}
          onChange={this.handleChange}
          list="purpose"
        />
        <datalist id="purpose">
          <option value="Vacation" />
          <option value="Business" />
          <option value="Other" />
        </datalist>
        <button type="button" className="submit-button" onClick={this.checkInputs}>Submit</button>
        { error ? <p>Please complete all fields!</p> : null }
      </form>
    );
  }
}

export default Form;

Form.propTypes = {
  addUser: PropTypes.func.isRequired,
};
