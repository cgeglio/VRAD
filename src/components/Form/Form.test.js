import React from 'react';
import { shallow } from 'enzyme';
import Form from './Form';

describe('Form', () => {
  let wrapper;
  const mockAddUser = jest.fn();

  beforeEach(() => {
    wrapper = shallow(<Form
      addUser={mockAddUser}
    />);
  });

  it('should match the snapshot with all data passed in correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should start off with a default state', () => {
    expect(wrapper.state()).toEqual({ name: '', email: '', purpose: '', error: false });
  });

  it('should update state with new user info when handleChange is called', () => {
    const mockEvent = { target: { name: 'name', value: 'Jeff' } };
    const expected = { name: 'Jeff', email: '', purpose: '', error: false };

    wrapper.instance().handleChange(mockEvent);
    expect(wrapper.state()).toEqual(expected);
  });

  it('should check that all inputs fields contain information when checkInputs is called', () => {
    const mockEvent = { preventDefault: jest.fn() };
    const mockState = { name: 'Jeff', email: 'jeff@turing.org', purpose: 'vacation', error: false };
    wrapper.instance().addUser = jest.fn();

    wrapper.setState(mockState);
    wrapper.instance().checkInputs(mockEvent);
    expect(wrapper.instance().addUser).toHaveBeenCalledWith(mockEvent);
  });

  it('should update state with an error any fields do not contain information when checkInputs is called', () => {
    const mockEvent = { preventDefault: jest.fn() };
    const expected = { name: '', email: '', purpose: '', error: true };
    wrapper.instance().addUser = jest.fn();

    wrapper.instance().checkInputs(mockEvent);
    expect(wrapper.state()).toEqual(expected);
  });


  it('should add a user when addUser is called and reset state', () => {
    const mockEvent = { preventDefault: jest.fn() };
    const mockState = { name: 'Jeff', email: 'jeff@turing.org', purpose: 'vacation', error: false };
    const expectedUser = { name: 'Jeff', email: 'jeff@turing.org', purpose: 'vacation' };
    const expectedState = { name: '', email: '', purpose: '', error: false };

    wrapper.setState(mockState);
    wrapper.instance().addUser(mockEvent);
    expect(mockAddUser).toHaveBeenCalledWith(expectedUser);
    expect(wrapper.state()).toEqual(expectedState);
  });

  it('should call the checkInputs method when the submit button is clicked', () => {
    wrapper.instance().checkInputs = jest.fn();
    wrapper.instance().forceUpdate();
    const mockEvent = { preventDefault: jest.fn() };

    wrapper.find('.submit-button').simulate('click', mockEvent);
    expect(wrapper.instance().checkInputs).toHaveBeenCalledWith(mockEvent);
  });
});
