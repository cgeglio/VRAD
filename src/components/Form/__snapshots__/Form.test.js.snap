// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should match the snapshot with all data passed in correctly 1`] = `
<form>
  <input
    name="name"
    onChange={[Function]}
    placeholder="Name..."
    type="text"
    value=""
  />
  <input
    name="email"
    onChange={[Function]}
    placeholder="Email..."
    type="text"
    value=""
  />
  <input
    list="purpose"
    name="purpose"
    onChange={[Function]}
    placeholder="Purpose..."
    type="text"
    value=""
  />
  <datalist
    id="purpose"
  >
    <option
      value="Vacation"
    />
    <option
      value="Business"
    />
    <option
      value="Other"
    />
  </datalist>
  <button
    className="submit-button"
    onClick={[Function]}
    type="button"
  >
    Submit
  </button>
</form>
`;
